File: LongestConsecutiveSequence/longest_consecutive_sequence.py
def longest_consecutive(nums: list[int]) -> int:
    if not nums:
        return 0
    sorted_nums = sorted(nums)
    longest = 1
    count = 1
    n0 = sorted_nums[0]
    for n in sorted_nums[1:]:
        if n == n0:
            continue
        if n - n0 == 1:
            count += 1
            n0 = n
        else:
            longest = max(longest, count)
            count = 1
            n0 = n
    return max(longest, count)

def longest_consecutive2(nums: list[int]) -> int:
    nums_set = set(nums)
    longest = 0
    for n in nums:
        if n - 1 in nums_set:
            continue
        count = 0
        while (n + count) in nums_set:
            count += 1
        longest = max(longest, count)
    return longest

File: LongestConsecutiveSequence/LongestConsecutiveSequence.hs
import Data.List (sort)
import Data.Set qualified as S

longestConsecutive :: [Int] -> Int
longestConsecutive [] = 0
longestConsecutive nums = max current longest
  where
    (current, longest, _) = foldl go (1, 1, head sorted) $ tail sorted
    sorted = sort nums
    go :: (Int, Int, Int) -> Int -> (Int, Int, Int)
    go (c, l, n0) n
      | n == n0 = (c, l, n)
      | n - n0 <= 1 = (c + 1, l, n)
      | otherwise = (1, max c l, n)

longestConsecutive2 :: [Int] -> Int
longestConsecutive2 [] = 0
longestConsecutive2 nums = foldr go 1 nums
  where
    numsSet = S.fromList nums
    sequenceLength :: Int -> Int -> Int
    sequenceLength c n
      | S.member (n + 1) numsSet = sequenceLength (c + 1) (n + 1)
      | otherwise = c
    go n a
      | S.member (n - 1) numsSet = a
      | otherwise = max a $ sequenceLength 1 n
